### PR TITLE
Can 97 -- V2

### PR DESCRIPTION
**Pull request by ahmed130544**

### CODE REVIEW

File: app/Http/Controllers/AdsController.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Http\Requests\AdsRequest;
+use App\Http\Resources\AdsResource;
+use App\Models\Page;
+use Illuminate\Http\Request;
```
Comment (akuks) [inline app/Http/Controllers/AdsController.php:8]: This we can remove if we are not using ut

File: app/Http/Controllers/ImageController.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Http\Requests\ImageRequest;
+use App\Http\Resources\ImageResource;
+use App\Models\Page;
+use Illuminate\Http\Request;
```
Comment (akuks) [inline app/Http/Controllers/ImageController.php:8]: We can remove it, if we are not using it

File: app/Http/Controllers/ImageController.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+class ImageController extends Controller
+{
+    /**
+     * Get the images list filtered by the specific page
+     * 
+     * @param ImageRequest $request
+     * @param Page $pageName
```
Comment (akuks) [inline app/Http/Controllers/ImageController.php:16]: Doesn't looks correct. There is only one param. Probably you need to use Swagger documentation. Please check other Controllers

File: app/Http/Controllers/AdsController.php [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+class AdsController extends Controller
+{
+    /**
+     * Get ads list filtered by the specific page 
+     * 
+     * @param AdRequest $request
+     * @param Page $pageName
```
Comment (akuks) [inline app/Http/Controllers/AdsController.php:16]: Doesn't looks correct. There is only one param. Probably you need to use Swagger documentation. Please check other Controllers

File: app/Http/Requests/AdsRequest.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
+<?php
```
Comment (akuks) [inline app/Http/Requests/AdsRequest.php:1]: There is already a Request. Please move it to there

File: app/Http/Requests/AdsRequest.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace App\Http\Requests;
+
+use Anik\Form\FormRequest;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Contracts\Validation\Validator;
```
Comment (akuks) [inline app/Http/Requests/AdsRequest.php:7]: I can't see the usage of it. Please check. And remove if we are not using it.

File: app/Http/Requests/AdsRequest.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+use Illuminate\Validation\ValidationException;
```
Comment (akuks) [inline app/Http/Requests/AdsRequest.php:8]: I can't see the usage of it. Please check. And remove if we are not using it.

File: app/Http/Requests/AdsRequest.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    protected function rules(): array
+    {
+        switch ($this->method()) {
```
Comment (akuks) [inline app/Http/Requests/AdsRequest.php:29]: For this, I don't think switch is good idea. We can replace it by `if`

File: app/Http/Requests/AdsRequest.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+            case 'POST': {
+                    return [
+                        'page_name' => 'required|string'
+                    ];
+                    break;
```
Comment (akuks) [inline app/Http/Requests/AdsRequest.php:35]: Probably needs to remove. This code is unreachable anyhow

File: app/Http/Requests/AdsRequest.php [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                break;
+        }
+    }
+
+    /**
+     * format the validation response if there is error in validation.
+     *
+     * @return array
+     */
+    protected function errorResponse(): ?JsonResponse
```
Comment (akuks) [inline app/Http/Requests/AdsRequest.php:47]: This can be replace with the `ResponseServiceProvider`

File: app/Http/Requests/ImageRequest.php [hunk 11]
```diff
@@ -0,0 +1,60 @@
+<?php
```
Comment (akuks) [inline app/Http/Requests/ImageRequest.php:1]: This should be under the `Request` folder. As it's already been present before

File: app/Http/Requests/ImageRequest.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace App\Http\Requests;
+
+use Anik\Form\FormRequest;
+use Illuminate\Http\JsonResponse;
+use Illuminate\Contracts\Validation\Validator;
```
Comment (akuks) [inline app/Http/Requests/ImageRequest.php:7]: Not used. We can remove it.

File: app/Http/Requests/ImageRequest.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+use Illuminate\Validation\ValidationException;
```
Comment (akuks) [inline app/Http/Requests/ImageRequest.php:8]: Not used. We can remove it.

File: app/Http/Requests/ImageRequest.php [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    protected function rules(): array
+    {
+        switch ($this->method()) {
```
Comment (akuks) [inline app/Http/Requests/ImageRequest.php:29]: Can be replaced by if()

File: app/Http/Requests/ImageRequest.php [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+            case 'POST': {
+                    return [
+                        'page_name' => 'required|string'
+                    ];
+                    break;
```
Comment (akuks) [inline app/Http/Requests/ImageRequest.php:35]: We can remove it. This code is not reachable anyhow

File: app/Http/Requests/ImageRequest.php [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * format the validation response if there is error in validation.
+     *
+     * @return array
+     */
+    protected function errorResponse(): ?JsonResponse
+    {
+
+        return response()->json([
```
Comment (akuks) [inline app/Http/Requests/ImageRequest.php:50]: Can be replaced with `ResponseSeriviceProvider`

File: app/Http/Resources/AdsResource.php [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+class AdsResource extends JsonResource
+{
+    /**
+     * Transform the resource into an array.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return array
+     */
+    public function toArray($request)
```
Comment (akuks) [inline app/Http/Resources/AdsResource.php:15]: probably `:array()` can be added.

Isn't it should be `public function toArray( Request $request)`?
I can see, in the comments it is the Request object.

File: app/Http/Resources/ImageResource.php [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+class ImageResource extends JsonResource
+{
+    /**
+     * Transform the resource into an array.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return array
+     */
+    public function toArray($request)
```
Comment (akuks) [inline app/Http/Resources/ImageResource.php:15]: probably :array() can be added.

Isn't it should be public function toArray( Request $request)?
I can see, in the comments it is the Request object.

Comment (akuks) [reply app/Http/Resources/ImageResource.php:15]: I think, it needs a refactor. We might need a common Resource class other than Success And Error Resource and then return all the resources from there. We can also use interface and  then extends its implementation.

File: database/migrations/2022_02_17_104703_create_pages_table.php [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+class CreatePagesTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        if (!Schema::hasTable('pages')) {
```
Comment (akuks) [inline database/migrations/2022_02_17_104703_create_pages_table.php:16]: Incorrect if statement I guess. I think it started to add after the countries table

File: database/migrations/2022_02_17_104703_create_pages_table.php [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+            Schema::create('pages', function (Blueprint $table) {
+                $table->id();
+                $table->string('name', 50)->unique();
+                $table->timestamps();
```
Comment (akuks) [inline database/migrations/2022_02_17_104703_create_pages_table.php:20]: Do we need this? We are using unixtime stamp across application

File: database/migrations/2022_02_17_104704_create_ads_table.php [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+class CreateAdsTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        if (!Schema::hasTable('ads')) {
```
Comment (akuks) [inline database/migrations/2022_02_17_104704_create_ads_table.php:16]: Incorrect if statement I guess. I think it started to add after the countries table

File: database/migrations/2022_02_17_104704_create_ads_table.php [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+            Schema::create('ads', function (Blueprint $table) {
+                $table->id();
+                $table->string('client_id', 50);
+                $table->unsignedBigInteger('page_id');
+                $table->string('slot', 50);
+                $table->string('format', '50')->nullable();
+                $table->tinyInteger('adtest')->default('0')->comment('0 => Off, 1 => On'); 
+                $table->tinyInteger('is_responsive')->default(1)->comment('0 => True, 1 => Off');
+                $table->tinyInteger('status')->default(1)->comment('0 => Inactive, 1 => Active');
+                $table->timestamps();
```
Comment (akuks) [inline database/migrations/2022_02_17_104704_create_ads_table.php:26]: Do we need this? We are using unix timestamps across application

File: database/migrations/2022_02_21_082239_create_images_table.php [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+    public function up()
+    {
+        Schema::create('images', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedBigInteger('page_id');
+            $table->string('title', 100);
+            $table->text('description');
+            $table->string('route', 100);
+            $table->string('url', 200);
+            $table->timestamps();
```
Comment (akuks) [inline database/migrations/2022_02_21_082239_create_images_table.php:23]: Do we need this? We are using unix timestamps across application.

File: routes/web.php [hunk 24]
```diff
@@ -54,4 +54,6 @@
         $router->post('add_nick_name','NicknameController@addNickName');
     });
 
+    $router->post('/pageAdsListing','AdsController@pageAdsListing');
```
Comment (akuks) [inline routes/web.php:57]: Please follow the correct url convention. It is not matching with what are using

File: routes/web.php [hunk 25]
```diff
@@ -54,4 +54,6 @@
...
+    $router->post('/pageImagesListing','ImageController@pageImagesListing');
```
Comment (akuks) [inline routes/web.php:58]: Please follow the correct url convention. It is not matching with what are using

**Review by akuks — CHANGES_REQUESTED** [review]
All good. Please do the requested changes. And push the code again.

File: database/migrations/2022_02_17_104704_create_ads_table.php (hunk 22 above)
Comment (ahmed130544) [reply database/migrations/2022_02_17_104704_create_ads_table.php:26]: I do not found of using Unix timestamp, anywhere in project, also looked into other migrations as well, but didn't able to find it. Can you please help?

File: app/Http/Requests/ImageRequest.php (hunk 16 above)
Comment (ahmed130544) [reply app/Http/Requests/ImageRequest.php:50]: Unable to update, package do now allow it due to compatibility issues, gives following exception

Declaration of App\Http\Request\AdsRequest::errorResponse(): ?App\Providers\ResponseServiceProvider must be
compatible with Anik\Form\FormRequest::errorResponse(): ?Illuminate\Http\JsonResponse (500 Internal Server
Error)

We are using package, we are using following package

https://github.com/ssi-anik/form-request

File: app/Http/Resources/AdsResource.php (hunk 17 above)
Comment (ahmed130544) [reply app/Http/Resources/AdsResource.php:15]: Unable to fix it, checked documentation but it's not working, can you please help.

https://laravel.com/api/5.8/Illuminate/Http/Resources/Json/JsonResource.html#method_toArray

image

image

File: app/Http/Request/AdsRequest.php [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * format the validation response if there is error in validation.
+     *
+     * @return array
+     */
+    protected function apiJsonResponse(): ?apiJsonResponse
+    {
+
+        return response()->json([
```
Comment (akuks) [inline app/Http/Request/AdsRequest.php:42]: Can we use this?
`return $this->resProvider->apiJsonResponse(422, $this->errorMessage(), '', $this->validator->errors()->messages());`

File: app/Http/Request/AdsRequest.php [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    /**
+     * format the validation response if there is error in validation.
+     *
+     * @return array
+     */
+    protected function apiJsonResponse(): ?apiJsonResponse
+    {
```
Comment (akuks) [inline app/Http/Request/AdsRequest.php:40]: We can also add a constructor

```javascript
public function __construct(ResponseInterface $resProvider)
    {
        $this->resProvider = $resProvider;
    }
```

File: app/Http/Request/ImageRequest.php [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * format the validation response if there is error in validation.
+     *
+     * @return array
+     */
+    protected function apiJsonResponse(): ?apiJsonResponse
+    {
+
+        return response()->json([
```
Comment (akuks) [inline app/Http/Request/ImageRequest.php:42]: Same as previous comments

**Review by akuks — CHANGES_REQUESTED** [review]
Please see the comments

**Review by akuks — APPROVED** [review]
All good

**Pull request merged**